aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+}
+
+// Init initializes the leader flush manager by enqueuing all
+// the flushers in the buckets.
+func (mgr *leaderFlushManager) Init(buckets []*flushBucket) {
+	mgr.Lock()
+	mgr.flushTimes.Reset()
+	for bucketIdx, bucket := range buckets {
+		mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	}
+	mgr.Unlock()
+}
+
+func (mgr *leaderFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.Duration) {
+	var (
+		shouldFlush = false
+		duration    tally.Timer
+		flushers    []PeriodicFlusher
+		waitFor     = mgr.checkEvery
+	)
+	mgr.Lock()
+	numFlushTimes := mgr.flushTimes.Len()
+	mgr.metrics.queueSize.Update(float64(numFlushTimes))
+	if numFlushTimes > 0 {
+		earliestFlush := mgr.flushTimes.Min()
+		if nowNanos := mgr.nowNanos(); nowNanos >= earliestFlush.timeNanos {
+			// NB(xichen): make a shallow copy of the flushers inside the lock
+			// and use the snapshot for flushing below.
+			shouldFlush = true
+			bucketIdx := earliestFlush.bucketIdx
+			duration = buckets[bucketIdx].duration
+			flushers = buckets[bucketIdx].flushers
+			nextFlushMetadata := flushMetadata{
+				timeNanos: earliestFlush.timeNanos + int64(buckets[bucketIdx].interval),
+				bucketIdx: bucketIdx,
+			}
+			mgr.flushTimes.Pop()
+			mgr.flushTimes.Push(nextFlushMetadata)
+			mgr.flushedSincePersist = true
+		} else {
+			// NB(xichen): don't oversleep if the next flush is about to happen.
+			timeToNextFlush := time.Duration(earliestFlush.timeNanos - nowNanos)
+			if timeToNextFlush < waitFor {
+				waitFor = timeToNextFlush
+			}
+		}
+	}
+
+	var (
+		now           = mgr.nowFn()
+		shouldPersist = false
+		flushTimes    schema.ShardSetFlushTimes
+	)
+	durationSinceLastPersist := now.Sub(mgr.lastPersistAt)
+	if mgr.flushedSincePersist && durationSinceLastPersist >= mgr.flushTimesPersistEvery {
+		shouldPersist = true
+		mgr.lastPersistAt = now
+		mgr.flushedSincePersist = false
+		flushTimes = mgr.prepareFlushTimesWithLock(buckets)
+	}
+	mgr.Unlock()
+
+	if !shouldFlush && !shouldPersist {
+		return nil, waitFor
+	}
+	mgr.flushTask.shouldFlush = shouldFlush
+	mgr.flushTask.duration = duration
+	mgr.flushTask.flushers = flushers
+	mgr.flushTask.shouldPersist = shouldPersist
+	mgr.flushTask.flushTimes = &flushTimes
+	return mgr.flushTask, 0
+}
+
+// NB(xichen): if the current instance is a leader, we need to update the flush
+// times heap for the flush goroutine to pick it up.
+func (mgr *leaderFlushManager) OnBucketAdded(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	mgr.Lock()
+	mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	mgr.Unlock()
+}
+
+// NB(xichen): leader flush manager can always lead.
+func (mgr *leaderFlushManager) CanLead() bool { return true }
+
+func (mgr *leaderFlushManager) enqueueBucketWithLock(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	nowNanos := mgr.nowNanos()
+	flushInterval := bucket.interval
+	nextFlushNanos := nowNanos + int64(flushInterval)
+	if mgr.jitterEnabled {
+		jitterNanos := int64(mgr.rand.Int63n(int64(flushInterval)))
```

Comment:
Yeah shouldn't we do a jitter max per bucket?  Otherwise if we have 1 hour tiles it could be a full 59 minutes after the the hour bucket is aggregated before we flush.

Even for a 10 minute bucket we could be 9 minutes late.  I should think we want this to be configurable per bucket.  Or at the very least some fixed % of the `flushInterval` size (say 10% - but even then this would likely want to be configurable per bucket size as you're probably ok for 1 second buckets to be 100% of the flush interval to be jittered).